android/WALT/app/src/main/java/org/chromium/latency/walt/MidiFragment.java

```diff
@@ -0,0 +1,285 @@
+    private TextView mTextView;
+
+    private MidiDevice mMidiDevice;
+    private MidiOutputPort mOutputPort;
+    private MidiInputPort mInputPort;
+
+    private long last_tWalt = 0;
+    private long last_tSys = 0;
+    private long last_tJava = 0;
+
+    private static final int noteDelay = 300;
+    private static final int timeout = 1000;
+
+    public MidiFragment() {
+        // Required empty public constructor
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+
+        activity = (MainActivity) getActivity();
+        logger = activity.logger;
+
+        findMidiDevice((MidiManager) container.getContext().getSystemService(Context.MIDI_SERVICE));
+
+        // Inflate the layout for this fragment
+        return inflater.inflate(R.layout.fragment_midi, container, false);
+    }
+
+    private void findMidiDevice(MidiManager m) {
+        MidiDeviceInfo[] infos = m.getDevices();
+        for(MidiDeviceInfo info : infos) {
+            String name = info.getProperties().getString(MidiDeviceInfo.PROPERTY_NAME);
+            logger.log("Found MIDI device named " + name);
+            if(TEENSY_MIDI_NAME.equals(name)) {
+                logger.log("^^^ using this device ^^^");
+                m.openDevice(info, new MidiManager.OnDeviceOpenedListener() {
+                            @Override
+                            public void onDeviceOpened(MidiDevice device) {
+                                if (device == null) {
+                                    logger.log("Error, unable to open MIDI device");
+                                } else {
+                                    logger.log("Opened MIDI device successfully!");
+                                    mMidiDevice = device;
+                                }
+                            }
+                        }, null);
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        mTextView = (TextView) activity.findViewById(R.id.txt_box_midi);
+
+        // Register this fragment class as the listener for some button clicks
+        activity.findViewById(R.id.button_start_midi_in).setOnClickListener(this);
+        activity.findViewById(R.id.button_start_midi_out).setOnClickListener(this);
+
+        // mLogTextView.setMovementMethod(new ScrollingMovementMethod());
+        mTextView.setText(activity.logger.getLogText());
+        activity.logger.broadcastManager.registerReceiver(mLogReceiver,
+                new IntentFilter(activity.logger.LOG_INTENT));
+
+    }
+
+    @Override
+    public void onPause() {
+        logger.broadcastManager.unregisterReceiver(mLogReceiver);
+        super.onPause();
+    }
+
+    @Override
+    public void onClick(View v) {
+        if(mMidiDevice == null) {
+            logger.log("Wait until the MIDI devices connects");
+        } else {
```

Comment:
Returning in the "if" and thus avoiding the need for "else" makes for a bit more readable code (and less unnecessary indentation).